[FLINK-22239][jdbc] Support exactly-once (sink) for PgSql and MySql

rkhachatryan opened the pull request:
## What is the purpose of the change

Implement workaround for limitation of some XA-compliant databases
of a single XA-transaction per connection by pooling connections.

Docs need to be updated but can be done separately in FLINK-22289.

## Verifying this change

Update existing `JdbcExactlyOnceSinkE2eTest` to work against (containerized) postgresql.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

curcur commented on flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/xa/XaFacadePoolingImpl.java:
```diff
@@ -0,0 +1,183 @@
+    // WARN: action MUST leave the facade in IDLE state (i.e. not start/end/prepare any tx)
+    private void runForXid(Xid xid, ThrowingConsumer<XaFacade, TransientXaException> action) {
+        XaFacade mapped = prepared.remove(xid);
+        if (mapped == null) {
+            LOG.debug("No XA resource found associated with XID: {}", xid);
+            action.accept(peekPooled());
```
when this would happen?

that says we have a prepared trans to commit/rollback/fail without a connection associated with it.

During the recovery I guess.

Document it a bit?

rkhachatryan replied:
Yes, this can happen on recovery, I'll add a comment.

curcur commented on flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/xa/XaFacade.java:
```diff
@@ -52,8 +51,8 @@
 
     /** @return a non-serializable instance. */
```
This is only for test?

Mark as visible for test?

curcur replied:
and other similar places.

rkhachatryan replied:
This particular method is used in `JdbcSink#exactlyOnceSink`.
The comment is misleading however; I'll remove it.

curcur commented on flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/xa/XaFacadeImpl.java:
```diff
@@ -82,12 +82,12 @@
 
     /** @return a non-serializable instance. */
```
same as above

curcur commented on flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/xa/JdbcXaSinkFunction.java:
```diff
@@ -286,12 +287,15 @@ private void prepareCurrentTx(long checkpointId) throws IOException {
         outputFormat.flush();
         try {
             xaFacade.endAndPrepare(currentXid);
+            outputFormat.reconnect(false);
```
Is this the fix for FLINK-22311?

rkhachatryan replied:
No, the intention here is to associate `outputFormat` with a new connection (I'll add a comment).

curcur commented:
Hey, @rkhachatryan , I've done with the review.

1. Overall the logic is pretty clear, please refer to a few of my comments.
2. The exactly-once end2end test is too simple, it only includes a test for insert, at least add one failure recovery test?
3. Address the connection closing issue found yesterday on top of "https://github.com/apache/flink/pull/15627" 
4. I think this PR might also fix the problem a user reported when connecting to Oracle, please double-check (FLINK-22311)

rkhachatryan commented on flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/xa/XaFacadePoolingImpl.java:
```diff
@@ -0,0 +1,183 @@
+    private final FacadeSupplier facadeSupplier;
+    private transient XaFacade active;
+    private transient Map<Xid, XaFacade> prepared;
+    private transient Deque<XaFacade> pooled;
```
`XaFacade` can be created by deserialization so these fields have to be initializated later and can't be made final unfortunately.

rkhachatryan commented on flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/xa/XaFacadePoolingImpl.java:
```diff
@@ -0,0 +1,183 @@
+    @Override
+    public void start(Xid xid) throws TransientXaException {
+        active.start(xid);
+    }
```
Very good point, I moved this logic from `endAndPrepare` to `start` (I also had to adjust `JdbcXaSinkFunction` to access `getConnection` only after `start`).

rkhachatryan commented:
Thanks for the review @curcur.

I addressed your comments and rebased the PR to master.
I also updated the end-to-end test and added a commit to rollback discovered "orphaned" transaction on recovery by default (initially this was a limitation but then I realized that the fix could be pretty simple).

The problem in FLINK-22311 is that `UnsupportedOperationException` is masking the actual error. That actual issue is now solved (re-connection it's still unsupported). I didn't succeed with Oracle, but I checked it manually with DB2, MySQL and PostgreSQL.

curcur commented on flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/xa/JdbcExactlyOnceSinkE2eTest.java:
```diff
@@ -102,19 +212,108 @@ public void cancel() {
         @Override
         public void notifyCheckpointComplete(long checkpointId) {
-            if (checkpointId == this.checkpointAfterData) {
-                dataCheckpointed = true;
+            if (checkpointId == this.lastCheckpointId) {
+                lastSnapshotConfirmed = true;
+            }
+        }
```
would this cause test in-stablity?

rkhachatryan replied:
I don't see any issue here, could you explain your concern?

curcur commented:
Thanks, @rkhachatryan for the fix. I am mostly fine with this PR with one-or-two places not completely understand. Let's sync up offline tomorrow.

Besides that, I think 
1. the code needs some clean-up. There are a few classes for example `JdbcXaFacadeTestHelper` having several methods unused.
2. the first and second commit does not self-compile because in
`JdbcXaSinkMigrationTest` and
`JdbcXaSinkTestBase`'s

`getDbMetadata().getUser()`,
`getDbMetadata().getPassword()`

methods not found.

rkhachatryan commented on flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/xa/XaFacadeImpl.java:
```diff
@@ -115,10 +116,6 @@ public void close() throws SQLException {
             connection.close();
             connection = null;
         }
-        if (xaConnection != null) {
-            xaConnection.close();
-            xaConnection = null;
-        }
```
Closing XA connection issues a rollback, so a prepared can be rolled back (or attempted).

curcur commented:
I've done with the tests as well.

curcur commented:
Thanks for fixing this @rkhachatryan !

rkhachatryan commented:
Thanks for the review @curcur.
CI failure unrelated (FLINK-22122), merging.
